### PR TITLE
livecheck/strategy/sparkle.rb: Add `macos` to the candidate `os` values list

### PR DESCRIPTION
**Pull request by timvisher**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

In attempting to [bump TextExpander to the latest available release](https://github.com/Homebrew/homebrew-cask/pull/134073) I found that the latest versions weren't being detected by livecheck. This is because due to [a prior change](https://github.com/Homebrew/brew/pull/10273) we (correctly) attempt to filter the OS down to just macOS. Unfortunately, the filter was a bit too aggressive. This adds the new observed value from the TextExpander Sparkle feed.

```
$ date -u '+%FT%T%z' && xmlstarlet sel -t -v '//rss//channel//item//enclosure/@*[name()="url" or name()="sparkle:version" or name()="sparkle:shortVersionString" or name()="sparkle:os"]' <(curl --location --silent https://textexpander.com/appcast/TextExpander-macOS.xml)
2022-10-22T17:07:06+0000
https://cdn.textexpander.com/mac/731.2/TextExpander_7.3.1.dmg
731.2
7.3.1
macos
https://cdn.textexpander.com/mac/720.16/TextExpander_7.2.dmg
720.16
7.2
macos
https://cdn.textexpander.com/mac/710.6/TextExpander_7.1.dmg
710.6
7.1
macos
https://cdn.textexpander.com/mac/702.2/TextExpander_7.0.2.dmg
702.2
7.0.2
macos
https://cdn.textexpander.com/mac/701.2/TextExpander_7.0.1.dmg
701.2
7.0.1
macos
https://cdn.textexpander.com/mac/700.33/TextExpander_7.0.dmg
700.33
7.0
macos
https://cdn.textexpander.com/mac/685.6/TextExpander_6.8.5.zip
685.6
6.8.5
https://cdn.textexpander.com/mac/684.8/TextExpander_6.8.4.zip
684.8
6.8.4
https://cdn.textexpander.com/mac/683.2/TextExpander_6.8.3.zip
683.2
6.8.3
https://cdn.textexpander.com/mac/682.10/TextExpander_6.8.2.zip
682.10
6.8.2
https://cdn.textexpander.com/mac/681.3/TextExpander_6.8.1.zip
681.3
6.8.1
https://cdn.textexpander.com/mac/680.30/TextExpander_6.8.zip
680.30
6.8
https://cdn.textexpander.com/mac/TextExpander_6.5.6.zip
656.3
6.5.6
https://cdn.textexpander.com/mac/TextExpander_6.5.5.zip
655.0
6.5.5
https://cdn.textexpander.com/mac/TextExpander_6.5.4.zip
654.3
6.5.4
https://cdn.textexpander.com/mac/TextExpander_6.5.3.zip
653.3
6.5.3
https://cdn.textexpander.com/mac/TextExpander_6.5.2.zip
652.0
6.5.2
https://smilesoftware.com/downloads/test/TextExpander_6.5.1.zip
651.5
6.5.1
```

According to the ['spec'](https://sparkle-project.org/documentation/api-reference/Classes/SUAppcastItem.html#/c:objc(cs)SUAppcastItem(py)osString), `osx` isn't actually one of the 'known potential values' but it also makes no specification of allowed values other than in the [Publishing → Alternate download locations for other operating systems Docs](https://sparkle-project.org/documentation/publishing/#alternate-download-locations-for-other-operating-systems) where it only lists `windows` and `linux` alongside an explicit invitation to 'feel free to add other OS names as needed'. Rather than try to be too clever about what we consider to macOS I figure listing things out according to observation is probably the best move.

~I have not written tests for this just yet on the off chance that the changes isn't welcome. If it is I'll add tests for it.~

Tests have been written and through the first round of review.

### CODE REVIEW

**Review by samford — COMMENTED** [review]
Makes sense to me. We already have a test for a non-matching `os` value, so we would just need tests with a matching `os`.

If you would prefer, I can tinker with the tests and push a commit (I originally wrote these, so I'm familiar with the setup). Otherwise, let me know if you have any questions along the way.

**Comment by timvisher** [comment]
Thanks for the tacit approval, @samford! :)

I'll see if I can't make rapid progress on the test in the next day or two and send any questions your way.

**Comment by timvisher** [comment]
@samford Made some progress here. I think the tests need to be refactored a bit because the positional reference and sorting issues between `items_from_content` and `versions_from_content` are getting fairly unwieldy. But please let me know if you think I'm off on completely the wrong track. :)

**Comment by timvisher** [comment]
OK @samford I think this is roughly in the place I think it should be. Let me know what you think. The test refactor is pretty extensive but I think they're clearer now and support the added use case.

**Comment by samford** [comment]
Sorry for the delay. I've been busy with other things the past few days but I will try to go through the test changes tonight (EDT) or tomorrow.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:194]: With the addition of this `Item` hash, the `items` variable name becomes a bit confusing. I think we should rename `version_items` to `items` (that feels more straightforward/understandable to me) and rename `items` to `item_arrays` (so it's a little more clear that the hash values are `Item` arrays). What do you think about this?

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:209]:
```suggestion
        # `#items_from_content` falls back to a default `pub_date` when one
        # isn't provided or can't be successfully parsed.
```

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:239]:
```suggestion
      ],
      appcast_sorted: [
        version_items[:version_123],
        version_items[:version_121],
        version_items[:version_120],
        version_items[:version_122],
      ],
```

Instead of creating a sorted array later in the tests, it probably makes more sense to define it here.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:290]:
```suggestion
```

At least to me, I don't think there's value in explaining that `#items_from_content` doesn't sort items. I think it's reasonable to expect that the items will simply be in the order they appear in the appcast. That said, I think we should omit this comment and simply explain how `#versions_from_content` sorts items before the related tests (as that's more likely to be unexpected behavior).

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:313]:
```suggestion
  # `#versions_from_content` sorts items by `pub_date` and `bundle_version`, so
  # these tests have to account for this behavior in the expected output.
  # For example, the version 122 item doesn't have a parseable `pub_date` and
  # the substituted default will cause it to be sorted last.
```

The existing comment may be a little confusing in places, so here's an alternative take that may be more straightforward.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:321]:
```suggestion
    let(:subbed_items) { items[:appcast_sorted].map { |item| item.nice_version.sub("1", "0") } }
```

With the addition of the `:appcast_sorted` array, this can return to the previous formatting. Past that, I don't think the additional commentary here is necessary/beneficial (i.e., one can simply look a few lines down to see how `subbed_items` is used).

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:332]:
```suggestion
```

Removing in favor of the aforementioned `:appcast_sorted` array.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:352]:
```suggestion
      ).to eq([subbed_items[0]])
```

Reverting in relation to the changes above.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:359]:
```suggestion
     ).to eq(subbed_items)
```

Reverting in relation to the changes above.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:403]:
```suggestion
      ).to eq(items[:appcast_sorted].map(&:short_version))
```

Updating in relation to the changes above.

**Review by samford — CHANGES_REQUESTED** [review]
I've gone through this and added comments and suggested changes. Outside of the areas I've highlighted, I tend to agree with the general idea behind some of the test changes. Adapting the existing structure to accommodate the necessary `os`-related additions was always going to be a little tricky, so thanks for engaging with the complexity.

I'll try to take another look at this when I'm awake again (it's very late over here), just in case I overlooked any areas or didn't explain things as well as I could [with a more well-rested mind].

**Comment by timvisher** [comment]
Thanks for going through all this, @samford. :)

I'll take a look at your comments in more detail when I have a moment. Hopefully over the next couple of days. But briefly skimming I don't see anything that should represent a large problem.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:11]: Sure. I feel like this was probably just `brew style --fix` doing its thing. :)

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:24]: I'm not sure whether your terse comment would have helped me or other new comers not be surprised quite so much. There's a lot of behavior that took me quite a bit to parse out that this comment describes. But for the sake of getting this change in I'll go ahead and accept it. Just figured I'd explain why it is the way it is.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:24]: Sure. Same thing as my other response. This was very surprising behavior to me and it's from a fairly subtle change in the data. I'll accept this but I think this comment would've helped me quite a bit getting up to speed on changing the tests.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:121]: I feel a little more strongly about this than I do about the other changes that I'm accepting.

I find the approach of just explicitly building up the string you expect rather than starting with the existing `appcast` string and substituting a substitution of `version_123_item` into it requires a lot of contextual awareness and introduces coupling between `appcast` and `beta_channel_item`.

I think it's significantly cleaner (but also slightly WETter) to just build up the `beta_channel_item` directly from the components. This decouples `appcast` and `beta_channel_item` and, to my eyes at least, reads much more clearly.

For now I'd like to keep this but please let me know if you still feel strongly that it should be changed.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:137]: This is exactly my point for adding these verbose explanatory comments. Before I made this change I _wasn't_ familiar with how that worked and it was quite surprising to me, especially the sort order differences. :)

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:135]: See my comment here https://github.com/Homebrew/brew/pull/14037/files#r1027104421

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:163]: I agree with this. I'll work that in.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:248]: Actually it looks like via the magick of Ruby I can reference the `appcast_url` before it's declared. 🤯

**Comment by timvisher** [comment]
👋 @samford. Anything I can do to help push this along? :)

**Comment by samford** [comment]
Thanks for the reminder. I'll try to take another pass through this later today when I have a free moment.

**Review by samford — COMMENTED** [review]
I took another pass through this and the only area that still gives me pause is the explicit  `beta_channel_item` and `undefined_namespace` approach. From my perspective, it increases our maintenance burden (having to carefully keep these strings in parity with `appcast` when any changes are made) and makes it a bit harder to understand what's different between these strings in comparison to `appcast` from a quick glance.

However, it's not all that different from how we approach the variables like `v123_item`, `v122_item`, etc. and, if you're any indicator, the explicit approach may be easier for some to understand. I don't think there's much value in further arguing my point, as we can always return to the substutition approach in the future if this explicit approach doesn't serve us in the long run. In the interim time, it's fine enough for me.

There's one small suggestion below but that should be it for changes here. Once we've resolved that, feel free to squash and rebase this on the latest `master` branch and I'll approve/merge it when I have a chance. Thanks again for all your work here and I appreciate your patience as we worked through this (I've been fairly busy lately).

**Comment by timvisher** [comment]
> if you're any indicator, the explicit approach may be easier for some to understand. I don't think there's much value in further arguing my point, as we can always return to the substutition approach in the future if this explicit approach doesn't serve us in the long run. In the interim time, it's fine enough for me.

Thanks very much for all the work on your part to get this reviewed and for being willing to give this approach a try for a bit. :)

I'll get things squashed up and re-tested and then we should be good to go!

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:126]: @samford I decided to go with the substitution approach here since I think it really is clearer. I don't love using a substitution to insert the beta channel into v123_item but I think it does make what `beta_channel_item` is supposed to be much clearer. 👍

Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:126]:
```suggestion
    beta_channel_item = appcast.sub(
      v123_item,
      v123_item.sub(
        "</title>",
        "</title>\n<sparkle:channel>beta</sparkle:channel>",
      ),
    )
```

I'm still in favor of inlining the value of `beta_channel_item_string`, since we don't reuse it (i.e., I don't think it's worth creating a variable to simplify the appearance of the code). If we don't like the nested multiline `#sub` calls, it's technically possible to have the `v123_item.sub` call on one line while still passing `brew style` (i.e., our line limit is 120 characters and this was just to make things a bit more readable for people like me who wrap at 80 characters).

Besides that, I slightly prefer anchoring the substitution to the closing `</title>` tag (as in the existing substitution), as that's common throughout all of the item XML examples in these tests.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:124]:
```suggestion
```

This goes along with the `beta_channel_item` suggestion and it's only separate because GitHub doesn't allow a suggestion across deleted lines.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [inline Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:140]:
```suggestion
    undefined_namespace = appcast.sub(/\s*xmlns:sparkle="[^"]+"/, "")
```

This is just a small tweak to reinstate part of the existing regex. Namely, we want to ensure that the substitution doesn't leave behind any leading whitespace. It *shouldn't* make a functional difference but better safe than sorry.

**Review by samford — COMMENTED** [review]
Just a couple tweaks in relation to the substitutions but this should be the last of it.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:126]: This is one of the substitutions that for whatever reason I just find hard to follow without the named intermediary. It's difficult for me to follow the double layer of substitutions and I don't think the named intermediary adds much noise for the sake of clarity.

That said I don't mind using the closing `</title>` as the anchor.

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:124]: See [my response](https://github.com/Homebrew/brew/pull/14037#discussion_r1038398834).

File: Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb
Comment (samford) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:127]: The way I think of it is "set `beta_channel_item` to a copy of `appcast` where the first item has `<sparkle:channel>beta</sparkle:channel>` appended after `</title>`". Any difficulty in reading it may arise from using a multiline `v123_item.sub` call within the multiline `appcast.sub` call, so I wouldn't be against the following alternative formatting if it helps to make it more understandable for you:

```suggestion
    # Set the first item in a copy of `appcast` to the "beta" channel, to test
    # filtering items by channel using a `strategy` block.
    beta_channel_item = appcast.sub(
      v123_item,
      v123_item.sub("</title>", "</title>\n<sparkle:channel>beta</sparkle:channel>"),
    )
```

With the added comment in this suggestion, I don't see this as something that's likely to trip up most readers. I still don't see the `beta_channel_item_string` change as potentially beneficial enough to warrant its inclusion, so I will have to pass on it. Hopefully this suggestion is more amenable to you but, either way, one of these two suggestions will need to be incorporated before this is merged.

Comment (timvisher) [reply Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb:127]: It's a shame we couldn't come to terms but I appreciate the effort you put in to reviewing things. I guess in the interest of getting the logical change in I'll go one of the suggestions and drop the discussion. :)

**Comment by samford** [comment]
Looks good to me at this point. If you don't mind squashing this again, I'll approve and merge once that's done.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @timvisher!

**Pull request merged**
